new.go

```diff
@@ -103,145 +104,162 @@ func newContainerIDMappingOptions(idmapOptions *IDMappingOptions) storage.IDMapp
 		}
+	}
 
-		destImage, err := localImageNameForReference(ctx, store, srcRef)
-		if err != nil {
-			return nil, "", nil, errors.Wrapf(err, "error computing local image name for %q", transports.ImageName(srcRef))
+	localImageRef, _, localImage, err := resolveLocalImage(systemContext, store, options)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	// If we could resolve the image locally, check if it was referenced by
+	// ID.  In that case, we don't need to bother any further and can
+	// prevent prompting the user.
+	if localImage != nil && strings.HasPrefix(localImage.ID, options.FromImage) {
+		return localImageRef, localImageRef.Transport().Name(), localImage, nil
+	}
+
+	if options.PullPolicy == PullNever || options.PullPolicy == PullIfMissing {
+		if localImage != nil {
+			return localImageRef, localImageRef.Transport().Name(), localImage, nil
 		}
-		if destImage == "" {
-			return nil, "", nil, errors.Errorf("error computing local image name for %q", transports.ImageName(srcRef))
+		if options.PullPolicy == PullNever {
+			return nil, "", nil, errors.Errorf("pull policy is %q but %q could not be found locally", "never", options.FromImage)
 		}
-		ref, err := is.Transport.ParseStoreReference(store, destImage)
+	}
+
+	// No transport, so let's use short-name resolution to find potential
+	// pullCandidates.
+	if len(pullCandidates) == 0 {
+		candidates, err := shortnames.Resolve(systemContext, options.FromImage)
 		if err != nil {
-			return nil, "", nil, errors.Wrapf(err, "error parsing reference to image %q", destImage)
+			return nil, "", nil, err
+		}
+		for i, candidate := range candidates {
+			ref, err := docker.ParseReference("//" + candidate.String())
+			if err != nil {
+				return nil, "", nil, err
+			}
+			pullCandidates = append(pullCandidates, dockerPullCandidate{ref: ref, alias: &candidates[i]})
 		}
+	}
 
-		if options.PullPolicy == PullIfNewer {
-			img, err := is.Transport.GetStoreImage(store, ref)
-			if err == nil {
-				// Let's see if this image is on the repository and if it's there
-				// then note it's Created date.
-				var repoImageCreated time.Time
-				repoImageFound := false
-				repoImage, err := srcRef.NewImage(ctx, systemContext)
-				if err == nil {
-					inspect, err := repoImage.Inspect(ctx)
-					if err == nil {
-						repoImageFound = true
-						repoImageCreated = *inspect.Created
-					}
-					repoImage.Close()
-				}
-				if !repoImageFound || repoImageCreated == img.Created {
-					// The image is only local or the same date is on the
-					// local and repo versions of the image, no need to pull.
-					return ref, transport, img, nil
-				}
+	var pullErrors []error
+	for _, pullCandidate := range pullCandidates {
+		name := pullCandidate.ref.StringWithinTransport()
+
+		// Skip this candidate if it's "older" than the local one.  If
+		// there's no local image, use the first successfully pulled
+		// remote one.
+		if options.PullPolicy == PullIfNewer && localImage != nil {
```

Comment:
Maybe it will be clearer further down the code, but why are we checking for a local image?  If we don't have a local image and the user has asked for a PullIfNewer, we should pull if the image is on the repository.  Do we now?

Reply:
Well testing seems to show it pulls the image, if appropriate to do, with this code.  I'll have to test further and think more on Tuesday.  Off on PTO Monday and probably won't get time.

Reply:
Thanks for checking!

In this case, when there's no matching local image, we will pull the first image we can find. We only need to do the comparison if the the policy is set to "newer" and if there's a local image. That's what the condition does. I'll add a more verbose comment to prevent further confusion.